packages/cardpay-sdk/sdk/claim-settlement-module.ts

```diff
@@ -19,39 +27,255 @@ export default class ClaimSettlementModule extends SafeModule {
   }
 
   async isValidState(claim: Claim, moduleAddress: string) {
     let module = new Contract(moduleAddress, this.abi, this.ethersProvider);
     return module.isValidState(claim.stateCheck.typeHash(), claim.stateCheck.abiEncode());
   }
 
-  async isValidCaller(claim: Claim, moduleAddress: string) {
+  async isValidCaller(claim: Claim, moduleAddress: string, caller: string) {
     let module = new Contract(moduleAddress, this.abi, this.ethersProvider);
-    return module.isValidCaller(claim.callerCheck.typeHash(), claim.callerCheck.abiEncode());
+    return module.isValidCaller(caller, claim.callerCheck.typeHash(), claim.callerCheck.abiEncode());
   }
 
   async isUsed(claim: Claim, moduleAddress: string) {
     let module = new Contract(moduleAddress, this.abi, this.ethersProvider);
     return module.used(claim.id);
   }
 
-  async hasBalance(claim: Claim, safeAddress: string) {
-    const action = claim.action.asMapping();
-    if (action.structName == 'TransferERC20ToCaller') {
+  async hasBalance(claim: Claim, moduleAddress: string): Promise<any> {
+    let { avatar: avatarAddress } = await this.getDetails(moduleAddress);
+    let action = claim.action.asMapping();
+    if (claim.action.structName == 'TransferERC20ToCaller') {
       let token = new Contract(action.token, ERC20ABI, this.ethersProvider);
-      let balance = await token.callStatic.balanceOf(safeAddress);
-      if (balance.gt(action.amount)) {
-        throw new Error('not enough balance');
+      let balance = await token.callStatic.balanceOf(avatarAddress);
+      if (balance.lt(action.amount)) {
+        return false;
       }
-    } else if (action.structName == 'TransferNFTToCaller') {
-      //TODO
+    } else if (claim.action.structName == 'TransferNFTToCaller') {
+      return false;
     } else {
-      throw new Error('action not implemented');
+      return false;
+    }
+    return true;
+  }
+
+  async addValidator(txnHash: string): Promise<SuccessfulTransactionReceipt>;
+  async addValidator(
+    moduleAddress: string,
+    avatarAddress: string,
+    validatorAddress: string,
+    txnOptions?: TransactionOptions,
+    contractOptions?: ContractOptions
+  ): Promise<SuccessfulTransactionReceipt>;
+  async addValidator(
+    moduleAddress: string,
+    avatarAddress?: string,
+    validatorAddress?: string,
+    txnOptions?: TransactionOptions,
+    contractOptions?: ContractOptions
+  ): Promise<SuccessfulTransactionReceipt> {
+    //TODO: Multi-signature. Only supports adding single validator only
+    if (!avatarAddress) {
+      throw new Error('avatarAddress must be specified');
+    }
+    if (!validatorAddress) {
+      throw new Error('validatorAddress must be specified');
+    }
+    let { nonce, onNonce, onTxnHash } = txnOptions ?? {};
+    let module = new Contract(moduleAddress, this.abi, this.ethersProvider);
+    let data = await module.interface.encodeFunctionData('addValidator', [validatorAddress]);
```

Comment:
So the validator for each claim module instance would be different, right?

Reply:
If you point to a new module instance (ie changing moduleAddress), it will have a new set of validators. Right now, we assume that there is only one avatar with one module instance attached.